include/triton/core/tritonserver.h

```diff
@@ -2655,11 +2656,14 @@ TRITONSERVER_MetricFamilyDelete(struct TRITONSERVER_MetricFamily* family);
 /// \param family The metric family to add this new metric to.
 /// \param labels The array of labels to associate with this new metric.
 /// \param label_count The number of labels.
+/// \param buckets Monotonically increasing values representing the
+/// bucket boundaries. For histogram only.
 /// \return a TRITONSERVER_Error indicating success or failure.
 TRITONSERVER_DECLSPEC struct TRITONSERVER_Error* TRITONSERVER_MetricNew(
     struct TRITONSERVER_Metric** metric,
     struct TRITONSERVER_MetricFamily* family,
-    const struct TRITONSERVER_Parameter** labels, const uint64_t label_count);
+    const struct TRITONSERVER_Parameter** labels, const uint64_t label_count,
+    const void* buckets = nullptr);
```

Comment:
We can not modify an existing C API function definition as it will break ABI compatibility. We need a new API instead, such as `TRITONSERVER_MetricNewV2` (name can be TBD).

Given this, the new C API should also aim to be more extensible to account for some of the shortcomings of the current `MetricNew` that necessitated this change.

For example, by introducing a configurable opaque object `TRITONSERVER_MetricSettings` that we can append new options as needed over time without breaking ABI compatibility:
```cpp
struct TRITONSERVER_MetricSettings;

// ...

TRITONSERVER_DECLSPEC struct TRITONSERVER_Error* TRITONSERVER_MetricNewV2(
    struct TRITONSERVER_Metric** metric,
    struct TRITONSERVER_MetricFamily* family,
    const struct TRITONSERVER_Parameter** labels, const uint64_t label_count,
    const struct TRITONSERVER_MetricSettings* settings);
```

`Histogram` will require `buckets` today.

`Summary` will require `quantiles`, and possibly some settings like `age` to account for update frequency, which we don't want to need yet another new C API for later on when we look at implementing this support.

So the `MetricSettings` object should be used to support adding new fields over time without breaking the API.

For easier maintenance, the C++ implementation of `TRITONSERVER_MetricNew` (for counter/gauge) internally can be updated to call `TRITONSERVER_MetricNewV2` with `settings = nullptr`.

Reply:
CC @GuanLuo @Tabrizian @oandreeva-nv for second opinion on new C API requirements

Reply:
I agree

Reply:
agree

Reply:
Agreed, one minor suggestion, can we name it `TRITONSERVER_MetricNewWithSettings` instead of v2? I think this is more consistent with what we've been doing with host policy etc.

Reply:
I named them `TRITONSERVER_MetricNewWithArgs` and `TRITONSERVER_MetricArgs` respectively before reading your comment. We can revise the namings if needed.

Reply:
Should be fine.